Fixes #13761 - libvirt provider with native bindings

lzap opened the pull request:
This new provider called _libvirt_ replaces _virsh_. It uses native libvirt
Ruby bindings, provides leases management and also contains full code coverage
with unit tests.

This PR replaces https://github.com/theforeman/smart-proxy/pull/382

We need new dependency in our repos - https://rubygems.org/gems/ruby-libvirt
and also we need to pull one extra fix to get leases reading:

http://libvirt.org/git/?p=ruby-libvirt.git;a=commit;h=c2d4192ebf28b8030b753b715a72f0cdf725d313

We can either wait for next release, include patch in OS packages or make
leases management optional, so users can use this.

lzap committed:
Fixes #13761 - libvirt provider with native bindings

lzap commented on bundler.d/libvirt.rb:
```diff
@@ -0,0 +1,6 @@
+  # version 0.6.0 or older won't work:
+  # http://libvirt.org/git/?p=ruby-libvirt.git;a=commit;h=c2d4192ebf28b8030b753b715a72f0cdf725d313
+  #gem 'ruby-libvirt', '> 0.6.0'
+  gem 'ruby-libvirt'
```
I asked upstream maintainer (Chris) about the next release...

dmitri-d commented on lib/proxy/libvirt_network.rb:
```diff
@@ -0,0 +1,72 @@
+      find_network.update command, section, -1, xml, flags
+    rescue ::Libvirt::Error => e
+      logger.error msg = "Error calling libvirt update: #{e}"
+      logger.debug xml if defined?(xml)
```
this will always evaluate to true -- xml is a method parameter.

dmitri-d commented on test/dhcp_libvirt/dhcp_libvirt_provider_test.rb:
```diff
@@ -0,0 +1,89 @@
+  def setup
+    @service = Proxy::DHCP::SubnetService.new
+    @subnet_store = @service.subnets = Proxy::MemoryStore.new
+    @subject = ::Proxy::DHCP::Libvirt::Provider.new.initialize_for_testing(
```
A bunch of tests fail here, as the constructor instantiates Proxy::LibvirtNetwork, which opens a connection to libvirt. Same issue in dns tests.

lzap commented:
Second attempt.

dmitri-d commented:
[test]

dmitri-d commented on lib/proxy/libvirt_network.rb:
```diff
@@ -0,0 +1,72 @@
+      network_update ::Libvirt::Network::UPDATE_COMMAND_ADD_LAST, ::Libvirt::Network::NETWORK_SECTION_IP_DHCP_HOST, xml
+    rescue ::Libvirt::Error => e
+      logger.error msg = "Error adding DHCP record: #{e}"
+      raise Proxy::DHCP::Error, msg
```
This (and other similar methods below) make core class depend on a module (::Proxy::DHCP::Error is defined in dhcp_common, ::Proxy::Dns::Error is defined in dns_common). One way to handle this would be to move the methods into subclasses in respective modules.

lzap replied:
Good observation, I moved the error handling into the "main" source files.

dmitri-d commented on modules/dhcp_libvirt/dhcp_libvirt_main.rb:
```diff
@@ -0,0 +1,100 @@
+    end
+
+    def initialize_for_testing(params)
+      initialize(params)
```
Not sure why you need to call this? `initialize_for_testing` is an instance method...

lzap replied:
Oh this was a leftover, removed.

dmitri-d commented on test/dhcp_libvirt/dhcp_libvirt_provider_test.rb:
```diff
@@ -0,0 +1,90 @@
+        "mac" => "52:54:00:13:05:12",
+        "expirytime" => 1_455_723_598
+    }]
+    @libvirt_network = mock()
```
You mock `::Libvirt.open` just so instantiation of `Proxy::DHCP::Libvirt` doesn't fail, and then you mock `::Proxy::LibvirtNetwork` object that internally uses `::Libvirt.open`. I think changing the constructor so it can accept an instance of `::Proxy::LibvirtNetwork` would require less mocks... Ditto for dns_libvirt_provider_test.

lzap replied:
Much better now, thanks.

dmitri-d commented:
Did you hear anything back from the ruby-libvirt maintainer?

lzap commented:
I haven't heard any response so far, no release yet. I fixed all your comments. I think the best thing to do is to test against stable version and make the error messages clear and merge it as is.

dmitri-d commented:
I'm getting "NULL pointer given" when trying to list all leases for a subnet. I haven't looked into this, but I suspect it's related to http://libvirt.org/git/?p=ruby-libvirt.git;a=commit;h=c2d4192ebf28b8030b753b715a72f0cdf725d313.  Dns add/remove record appears to be working ok.

lzap commented:
I have reproduced this with 0.6.0, that's actually what's being fixed in that patch. I created a workaround catching argumenterror exception and returning empty list of leases. Also fixed tests.

lzap commented on lib/proxy/libvirt_network.rb:
```diff
@@ -0,0 +1,63 @@
+    rescue ArgumentError => e
+      # workaround for ruby-libvirt < 0.6.1 (http://libvirt.org/git/?p=ruby-libvirt.git;a=commit;h=c2d4192ebf28b8030b753b715a72f0cdf725d313)
+      []
+    end
```
Here it is.

dmitri-d commented:
Rubocop is still failing.

lzap commented:
> Rubocop is still failing.

Oh thanks fixed.

## 

Later,
 Lukas #lzap Zapletal

dmitri-d commented:
This works ok with current libvirt. Sorry I didn't notice earlier:
- could you add a migration for libvirt 'network' parameter
- in https://github.com/theforeman/smart-proxy/blob/develop/config/settings.d/dhcp.yml.example: dhcp_virsh needs to be replaced with dhcp_libvirt
- in https://github.com/theforeman/smart-proxy/blob/develop/config/settings.d/dns.yml.example: dns_virsh needs to be replaced with dns_libvirt
- can you add a test similar to test/dhcp_isc/dhcp_isc_provider_interface_test.rb for dhcp_libvirt?
- dns_virsh and dhcp_virsh providers and their tests can be removed
-  lib/proxy/virsh.rb can be removed

lzap commented:
Added migration. Do you mean to add similar test for dns_libvirt? Because DHCP is already present.

I will remove virsh in separate commit on top of the current one once you don't have more comments for better clarity.

dmitri-d commented on lib/proxy/libvirt_network.rb:
```diff
@@ -0,0 +1,64 @@
+    end
+
+    def del_dhcp_record(record)
+      # ??? record[:hostname]
```
Did you forget to remove this comment?

dmitri-d commented:
The migration creates `dhcp_libvirt.yml` and `dns_libvirt.yml` config files and populates them with correct content. It doesn't copy the rest of config files however. Might be useful to add tests for the migration too? 

The rest looks ok.

lzap commented:
The rest of config files? I don't get it. Do you mean this:

```
    copy_original_configuration_except(path("dhcp_virsh.yaml", "dns_virsh.yml"))
```

dmitri-d commented:
Main config file and config files for other modules. Try executing your migration and you'll see that a bunch of files are missing.

lzap commented:
Added, also muted the annoying output during teste:

``` diff
diff --git a/extra/migrate_settings.rb b/extra/migrate_settings.rb
index 6a49b30..fd57ba8 100644
--- a/extra/migrate_settings.rb
+++ b/extra/migrate_settings.rb
@@ -132,10 +132,10 @@ module ::Proxy
       persist_migrations_state(executed_migrations, result_dir_path)
     end

-    def execute_migrations(migrations)
+    def execute_migrations(migrations, print_name_stdout = true)
       migrations.each do |migration|
         m = migration.new(working_dir_path)
-        puts m.migration_name.to_s
+        puts(m.migration_name.to_s) if print_name_stdout

         m.create_migration_dirs
         if migration == migrations.first
diff --git a/extra/migrations/20160411000000_migrate_libvirt_settings.rb b/extra/migrations/20160411000000_migrate_libvirt_settings.rb
index 7256050..b946233 100644
--- a/extra/migrations/20160411000000_migrate_libvirt_settings.rb
+++ b/extra/migrations/20160411000000_migrate_libvirt_settings.rb
@@ -4,6 +4,7 @@ require 'yaml'
 class MigrateVirshToLibvirtConfig < ::Proxy::Migration
   def migrate
     input_yaml = YAML.load_file(File.join(src_dir, 'settings.yml'))
+    copy_original_configuration_except(path("dhcp_virsh.yaml", "dns_virsh.yml"))
     write_yaml("dhcp_libvirt.yml", transform_dhcp_yaml(input_yaml))
     write_yaml("dns_libvirt.yml", transform_dns_yaml(input_yaml))
   end
diff --git a/lib/proxy/libvirt_network.rb b/lib/proxy/libvirt_network.rb
index 730f47f..bb51b66 100644
--- a/lib/proxy/libvirt_network.rb
+++ b/lib/proxy/libvirt_network.rb
@@ -29,7 +29,6 @@ module Proxy
     end

     def del_dhcp_record(record)
-      # ??? record[:hostname]
       nametag = "name=\"#{record.name}\"" if record.name
       xml = "<host mac=\"#{record.mac}\" ip=\"#{record.ip}\" #{nametag}/>"
       network_update ::Libvirt::Network::UPDATE_COMMAND_DELETE, ::Libvirt::Network::NETWORK_SECTION_IP_DHCP_HOST, xml
diff --git a/test/migrations/libvirt_migration_test.rb b/test/migrations/libvirt_migration_test.rb
new file mode 100644
index 0000000..502e6bb
--- /dev/null
+++ b/test/migrations/libvirt_migration_test.rb
@@ -0,0 +1,30 @@
+require 'test_helper'
+require File.join(File.dirname(__FILE__),'../../extra/migrate_settings')
+::Proxy::Migration.inject_migrations_instance(::Proxy::Migrations.new("dummy"))
+require File.join(File.dirname(__FILE__),'../../extra/migrations/20160411000000_migrate_libvirt_settings')
+
+class ProxyLibvirtMigrationTest < Test::Unit::TestCase
+
+  def setup
+    @old_config = YAML.load_file(File.join(File.dirname(__FILE__),'./migration_settings.yml'))
+    @migration = MigrateVirshToLibvirtConfig.new("/tmp")
+    @dhcp_data = @migration.transform_dhcp_yaml(@old_config.dup)
+    @dns_data = @migration.transform_dns_yaml(@old_config.dup)
+  end
+
+  def test_transform_dhcp_yaml_empty
+    assert_equal 'default', @migration.transform_dhcp_yaml({})[:network]
+  end
+
+  def test_output_has_correct_dhcp_network
+    assert_equal 'mynetwork', @dhcp_data[:network]
+  end
+
+  def test_transform_dns_yaml_empty
+    assert_equal 'default', @migration.transform_dns_yaml({})[:network]
+  end
+
+  def test_output_has_correct_dns_network
+    assert_equal 'mynetwork', @dns_data[:network]
+  end
+end
diff --git a/test/migrations/migration_settings.yml b/test/migrations/migration_settings.yml
index 0850c41..f1b946f 100644
--- a/test/migrations/migration_settings.yml
+++ b/test/migrations/migration_settings.yml
@@ -36,6 +36,8 @@
 :bmc: true
 :bmc_default_provider: ipmitool

+:virsh_network: mynetwork
+
 :chefproxy: true
 :chef_authenticate_nodes: true
 :chef_server_url: "https://foreman.example.com"
diff --git a/test/migrations/migration_test.rb b/test/migrations/migration_test.rb
index 87727dd..2fda6b4 100644
--- a/test/migrations/migration_test.rb
+++ b/test/migrations/migration_test.rb
@@ -101,7 +101,7 @@ class MigrationTest < Test::Unit::TestCase
     @migration.migrations.first.any_instance.expects(:create_migration_dirs)
     migrator.expects(:copy_original_configuration).with(migration.src_dir)

-    migrator.execute_migrations([@migration.migrations.first])
+    migrator.execute_migrations([@migration.migrations.first], false)

     assert_equal 1, migrator.executed_migrations.size
     assert migrator.executed_migrations.first.instance_of?(@migration.migrations.first)
diff --git a/test/migrations/monolithic_config_file_migration_test.rb b/test/migrations/monolithic_config_file_migration_test.rb
index f3417c8..a03a98b 100644
--- a/test/migrations/monolithic_config_file_migration_test.rb
+++ b/test/migrations/monolithic_config_file_migration_test.rb
@@ -24,7 +24,8 @@ class MonolithicConfigMigrationTest < Test::Unit::TestCase
                  :daemon_pid      => "/var/run/foreman-proxy/foreman-proxy.pid",
                  :log_file        => "/var/log/foreman-proxy/proxy.log",
                  :log_level       => "DEBUG",
-                 :https_port      => 8443
+                 :https_port      => 8443,
+                 :virsh_network   => "mynetwork"
   end

   def test_output_has_correct_tftp_settings
```

lzap commented:
I am not sure if this is expected or not:

```
[lzap@lzapx smart-proxy]$ ruby extra/migrate_settings.rb 
using:
  config file: /home/lzap/work/smart-proxy/config/settings.yml,
  modules config dir: /home/lzap/work/smart-proxy/config/settings.d,
  working dir: /home/lzap/work/smart-proxy/tmp,
  migrations dir: /home/lzap/work/smart-proxy/extra/migrations,
  migrations state path: /home/lzap/work/smart-proxy/config/migration_state

Running migrations...
migrate_monolithic_config
migrate_dns_settings
migrate_dhcp_settings
migrate_virsh_to_libvirt_config
Writing result to /home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/dhcp_libvirt.yml
Writing result to /home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/dns_libvirt.yml

[lzap@lzapx smart-proxy]$ find /home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/puppetca.yml.example
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/templates.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/logs.yml.example
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/pulp.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/bmc.yml.example
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/dns_virsh.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/dns_libvirt.yml.example
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/dhcp.yml.example
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/dhcp_isc.yml.example
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/dhcp_ms_native.yml.example
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/chef.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/dns_virsh.yml.example
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/dhcp_libvirt.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/puppet.yml.example
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/puppet.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/logs.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/dhcp_isc.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/discovery_image.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/colly.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/templates.yml.example
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/dns.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/dns_libvirt.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/dns.yml.example
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/dns_nsupdate.yml.example
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/tftp.yml.example
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/abrt.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/dhcp.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/dns_dnscmd.yml.example
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/tail.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/puppetca.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/realm.yml.example
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/dns_nsupdate_gss.yml.example
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/realm.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/bmc.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/facts.yml.example
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/dhcp_virsh.yml.example
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/dhcp_libvirt.yml.example
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/salt.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/discovery.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.d/tftp.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/dst/settings.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.d
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.d/puppetca.yml.example
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.d/templates.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.d/logs.yml.example
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.d/pulp.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.d/bmc.yml.example
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.d/dns_virsh.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.d/dns_libvirt.yml.example
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.d/dhcp.yml.example
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.d/dhcp_isc.yml.example
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.d/dhcp_ms_native.yml.example
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.d/chef.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.d/dns_virsh.yml.example
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.d/puppet.yml.example
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.d/puppet.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.d/logs.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.d/dhcp_isc.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.d/discovery_image.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.d/colly.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.d/templates.yml.example
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.d/dns.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.d/dns.yml.example
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.d/dns_nsupdate.yml.example
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.d/tftp.yml.example
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.d/abrt.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.d/dhcp.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.d/dns_dnscmd.yml.example
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.d/tail.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.d/puppetca.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.d/realm.yml.example
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.d/dns_nsupdate_gss.yml.example
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.d/realm.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.d/bmc.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.d/facts.yml.example
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.d/dhcp_virsh.yml.example
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.d/dhcp_libvirt.yml.example
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.d/salt.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.d/discovery.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.d/tftp.yml
/home/lzap/work/smart-proxy/tmp/migrate_virsh_to_libvirt_config/src/settings.yml
```

lzap commented:
Specifically, I am not sure if the temp file named `migrate_virsh_to_libvirt_config` is correct.

dmitri-d commented:
Results of each migration are kept in a dedicated directory named after the migration. You can use `rake migrate_settings` if you want migrations to replace your current configuration.

dmitri-d commented on extra/migrations/20160411000000_migrate_libvirt_settings.rb:
```diff
@@ -0,0 +1,29 @@
+class MigrateVirshToLibvirtConfig < ::Proxy::Migration
+  def migrate
+    input_yaml = YAML.load_file(File.join(src_dir, 'settings.yml'))
+    copy_original_configuration_except(path("dhcp_virsh.yaml", "dns_virsh.yml"))
```
This doesn't look right - `path` merges individual components into a filepath: https://github.com/theforeman/smart-proxy/blob/develop/extra/migrate_settings.rb#L72. You should use it on the line above though.

lzap replied:
Oh good catch.

I still plan to remove virsh provider in another commit once this is done.

dmitri-d commented:
The migration didn't update the main settings file: virsh_network setting is still present after migrations were run. Also, I wonder if old setting files (dns_virsh.yml and dhcp_virsh.yml) should be deleted too.

dmitri-d commented:
Also, could you rebase your branch against HEAD of develop?

lzap commented:
Rebased, fixed the migration, I believe now it is correct.

```
[lzap@lzapx smart-proxy]$ ruby extra/migrate_settings.rb 
using:
  config file: /home/lzap/work/smart-proxy/config/settings.yml,
  modules config dir: /home/lzap/work/smart-proxy/config/settings.d,
  working dir: /home/lzap/work/smart-proxy/tmp,
  migrations dir: /home/lzap/work/smart-proxy/extra/migrations,
  migrations state path: /home/lzap/work/smart-proxy/config/migration_state

Running migrations...
migrate_monolithic_config
migrate_dns_settings
migrate_dhcp_settings
migrate_virsh_to_libvirt_config
[lzap@lzapx smart-proxy]$ cat tmp/migrate_virsh_to_libvirt_config/dst/settings.d/*_libvirt.yml
---
:network: blah
---
:network: blah
[lzap@lzapx smart-proxy]$ grep blah tmp/migrate_virsh_to_libvirt_config/dst/settings.yml 
[lzap@lzapx smart-proxy]$ cat tmp/migrate_virsh_to_libvirt_config/dst/settings.d/*_virsh.yml
cat: tmp/migrate_virsh_to_libvirt_config/dst/settings.d/*_virsh.yml: No such file or directory
```

dmitri-d commented:
@lzap: this looks good -- wanna drop the old virsh stuff? I'll test it again and merge after that...

lzap committed:
Refs #13761 - removed legacy virsh provider

lzap commented:
Done in a separate commit, there are few references of the "virsh" term but all in migrations:

```
Searching 426 files for "virsh"

/home/lzap/work/smart-proxy/extra/migrations/20150327000000_migrate_monolithic_config.rb:
   36          :ssl_private_key    => :settings,
   37          :trusted_hosts      => :settings,
   38:         :virsh_network      => :settings,
   39          :foreman_url        => :settings,
   40          :settings_directory => :settings,

/home/lzap/work/smart-proxy/extra/migrations/20150611000000_migrate_dns_settings.rb:
   56  
   57    def recognized_dns_provider_name?(aname)
   58:     ['nsupdate', 'nsupdate_gss', 'virsh', 'dnscmd'].include?(aname)
   59    end
   60  

/home/lzap/work/smart-proxy/extra/migrations/20150826000000_migrate_dhcp_settings.rb:
   35  
   36    def old_provider_name_to_new(aname)
   37:     if ['isc', 'native_ms', 'virsh'].include?(aname)
   38        'dhcp_' + aname
   39      else

/home/lzap/work/smart-proxy/extra/migrations/20160411000000_migrate_libvirt_settings.rb:
    2  require 'yaml'
    3  
    4: class MigrateVirshToLibvirtConfig < ::Proxy::Migration
    5    def migrate
    6      input_yaml = YAML.load_file(path(src_dir, "settings.yml"))
    7:     copy_original_configuration_except("settings.yml", path("settings.d", "dhcp_virsh.yml"), path("settings.d", "dns_virsh.yml"))
    8      write_yaml(path(dst_dir, "settings.d", "dhcp_libvirt.yml"), transform_dhcp_yaml(input_yaml))
    9      write_yaml(path(dst_dir, "settings.d", "dns_libvirt.yml"), transform_dns_yaml(input_yaml))
   ..
   12  
   13    def transform_settings_yaml(yaml)
   14:     yaml.delete(:virsh_network)
   15      yaml
   16    end
   17  
   18    def transform_dns_yaml(yaml)
   19:     network = yaml[:virsh_network] || 'default'
   20      { :network => network }
   21    end
   22  
   23    def transform_dhcp_yaml(yaml)
   24:     network = yaml[:virsh_network] || 'default'
   25      { :network => network }
   26    end

/home/lzap/work/smart-proxy/test/migrations/dhcp_migration_test.rb:
   13      assert_equal 'dhcp_isc', @migration.old_provider_name_to_new('isc')
   14      assert_equal 'dhcp_native_ms', @migration.old_provider_name_to_new('native_ms')
   15:     assert_equal 'dhcp_virsh', @migration.old_provider_name_to_new('virsh')
   16      assert_equal 'unknown', @migration.old_provider_name_to_new('unknown')
   17    end

/home/lzap/work/smart-proxy/test/migrations/dns_migration_test.rb:
   16      assert_equal 'dns_nsupdate_gss', @migration.migrate_dns_provider_name('nsupdate_gss')
   17      assert_equal 'dns_dnscmd', @migration.migrate_dns_provider_name('dnscmd')
   18:     assert_equal 'dns_virsh', @migration.migrate_dns_provider_name('virsh')
   19      assert_equal 'blah', @migration.migrate_dns_provider_name('blah')
   20    end

/home/lzap/work/smart-proxy/test/migrations/libvirt_migration_test.rb:
    8    def setup
    9      @old_config = YAML.load_file(File.join(File.dirname(__FILE__),'./migration_settings.yml'))
   10:     @migration = MigrateVirshToLibvirtConfig.new("/tmp")
   11      @dhcp_data = @migration.transform_dhcp_yaml(@old_config.dup)
   12      @dns_data = @migration.transform_dns_yaml(@old_config.dup)

/home/lzap/work/smart-proxy/test/migrations/migration_settings.yml:
   37  :bmc_default_provider: ipmitool
   38  
   39: :virsh_network: mynetwork
   40  
   41  :chefproxy: true

/home/lzap/work/smart-proxy/test/migrations/monolithic_config_file_migration_test.rb:
   26                   :log_level       => "DEBUG",
   27                   :https_port      => 8443,
   28:                  :virsh_network   => "mynetwork"
   29    end
   30  

16 matches across 10 files
```

dmitri-d commented:
`/test/dhcp_libvirt/dhcp_libvirt_provider_test.rb` and `test/dns_libvirt/dns_libvirt_provider_test.rb` fail if corresponding settings files contain network name other than 'default': `:network => 'default'` isn't being passed to the constructor in these tests. 

Also, could you remove all comments from `test/migrations/migration_dns_settings.yml` -- no point in having the list of providers there, and replacing virsh to libvirt only makes them confusing.

lzap commented:
Do you mean this test:

``` ruby
  def test_default_settings
    assert_equal 'default', Proxy::DHCP::Libvirt::Provider.new(:libvirt_network => @libvirt_network).network
  end
```

?

lzap commented:
I thought it's the purpose of this test to test for the default value. I copied this from existing tests.

dmitri-d commented:
Right, so passing a parameter into the constructor will defeat the purpose of those test. You can override settings in the config file by calling `::Proxy::DHCP::Libvirt::Plugin.load_test_settings({})` -- should do the trick.

lzap commented:
Done, squashed the two changes into the two commits.

dmitri-d commented:
... and a couple more things (sorry!). 
- dns api now supports IPv6; I just checked and in libvirt's case it would only require something like (+ renaming methods in ::Proxy::LibvirtNetwork + tests). I think it makes sense to this right away, as opposed to creating noop methods expected by the interface now and actual implementation later.

```
  def create_aaaa_record(fqdn, ip)
    create_a_record(fqdn, ip)
  end

  def remove_aaaa_record(fqdn)
    remove_a_record(fqdn)
  end
```
- Could you move dns- and dhcp-specific code from `lib/proxy/libvirt_network.rb` into corresponding modules?

lzap commented:
Done, added IPv6 support, split of the proxy class into libvirt_common and submodules as subclasses.

dmitri-d commented on modules/dhcp_libvirt/dhcp_libvirt_main.rb:
```diff
@@ -0,0 +1,105 @@
+
+    def initialize(options = {})
+      @network = options[:network] || Proxy::DHCP::Libvirt::Plugin.settings.network
+      @libvirt_network = options[:libvirt_network] || ::Proxy::LibvirtNetwork.new(
```
Should be ::Proxy::LibvirtDHCPNetwork. Also: I know namespaces aren't consistent in DHCP module, but I'm trying to move everything into ::Proxy::DHCP (::Proxy::DHCP::Libvirt in your case) module, could you move LibvirtDHCPNetwork into ::Proxy::DHCP::Libvirt module?

dmitri-d commented on modules/dns_libvirt/dns_libvirt_main.rb:
```diff
@@ -0,0 +1,60 @@
+
+    def initialize(options = {})
+      @network = options[:network] || Proxy::Dns::Libvirt::Plugin.settings.network
+      @libvirt_network = options[:libvirt_network] || ::Proxy::LibvirtNetwork.new(
```
Should be ::Proxy::LibvirtDNSNetwork. Same comment about namespace as for dhcp_libvirt_main.rb.

lzap replied:
Good catch, will do.

dmitri-d commented on test/dhcp_libvirt/dhcp_libvirt_provider_test.rb:
```diff
@@ -0,0 +1,86 @@
+    @subject.initialize_for_testing(:service => @service)
+  end
+
+  def test_default_settings
```
Probably worth adding a test verifying that the correct network class was used to initialize the provider?

lzap replied:
Yup.

dmitri-d commented on test/dns_libvirt/dns_libvirt_provider_test.rb:
```diff
@@ -0,0 +1,76 @@
+
+  def test_provider_initialization
+    ::Proxy::Dns::Libvirt::Plugin.load_test_settings(:network => 'some_network')
+    assert_equal "some_network", Proxy::Dns::Libvirt::Record.new(:libvirt_network => @libvirt_network).network
```
Ditto here, a test verifying that the correct network class was used to initialize the provider would be useful.

lzap commented:
Rebased.

lzap commented:
Fixed tests, libvirt connection is now lazy-opened.

dmitri-d commented:
Merged as 6798dc5. Thanks, @lzap!
